File: actividades/prog/ejemplos/holamundos/j-canarias/src/aplicaciones/terminal/shell/ShellExec.java
package terminal.shell;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.File;

/**
 * <p>Esta clase implementa la misma funcionalidad que una consola
 * o interfaz de modo comando.</p>
 * <p>Actúa de intermediario entre elusuario y la auténtica shell del sistema.
 * Intercepta los mensajes entre ambos niveles.</p>
 * 
 * @author David Vargas <dvargas@canarias.org>
 */
public class ShellExec implements InterfazShell {

    private boolean fin;
    private String salida;
    private String prompt;
    private File pwd;
    private BufferedReader in;
    private Runtime runtime;

    /**
     * Constructor de la clase ShellAPIjava
     */
    public ShellExec() {
        fin = false;
        pwd = new File("/home/david/tmp");
        prompt = "[" + pwd + "]$ ";
        salida = "Consola ShellExec 0.1.0\n" + "Fecha: 2007-12-29\n" + "Autor: David Vargas\n";
        salida = salida + prompt;
        runtime = Runtime.getRuntime();
    }

    public String getPrompt() {
        return prompt;
    }

    public boolean isFin() {
        return this.fin;
    }

    public void setEntrada(String cmd) {
        cmd = cmd.trim();
        if (cmd.equals("exit")) {
            //Salir de la consola
            this.fin = true;
            this.salida = "Saliendo de la consola\n";
        } else {
            String s = "", t = "";
            try {
                Process p = runtime.exec(cmd);
                p.waitFor(); // wait for process to complete
                in = new BufferedReader(new InputStreamReader(p.getInputStream()));
                while ((t = in.readLine()) != null) {
                    s = s + t + "\n";
                }
            } catch (Exception e) {
                System.err.println("Error: " + e);
            }
            salida = s + "\n" + prompt;
        }
    }

    public String getSalida() {
        if (fin) {
            salida = "";
        }
        return salida;
    }
}

File: actividades/programar/files/ruby/mil-ejemplos/aplicaciones/terminal/clases/shell/shell_exec.rb
=begin
Fichero: clases/shell/ShellExec
Autor: David VArgas <dvargas@canarias.org>
Versión: 0.1.0 20080123
Descripción: Implementa una consola pasando los comandos
directamente al sistema.
=end
 

class ShellExec
  attr_reader :directorio, :prompt, :salida
  
  def initialize(*args)
    if (args.size>0)
      @directorio = args[0]
      ejecutar "cd "+@directorio
    else
      @directorio = "/home/david/tmp"
    end
    @fin = false
  end
  
  def prompt
    return "Consola Virtual ["+@directorio+"]$ "
  end
  
  def ejecutar(comando)
    if (comando=="exit")
      @fin = true
      @salida = "Saliendo de la consola virtual..."
    else
      orden = "`"+comando+"`"
      @salida = eval orden
      #@salida = @salida + prompt
    end
  end
  
  def fin?
    return @fin
  end
end
